In [3]:
from sklearn.datasets import load_wine
import pandas as pd

In [4]:
wine = load_wine()
df = pd.DataFrame(wine.data, columns = wine.feature_names)
df["class"] = wine.target

display(df.head())

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,class
0,14.23,1.71,2.43,15.6,127.0,2.80,3.06,0.28,2.29,5.64,1.04,3.92,1065.0,0
1,13.20,1.78,2.14,11.2,100.0,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050.0,0
2,13.16,2.36,2.67,18.6,101.0,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185.0,0
3,14.37,1.95,2.50,16.8,113.0,3.85,3.49,0.24,2.18,7.80,0.86,3.45,1480.0,0
4,13.24,2.59,2.87,21.0,118.0,2.80,2.69,0.39,1.82,4.32,1.04,2.93,735.0,0


### SIS1 Step 2: Inspect the Dataset

1. Print the shape of the dataset

In [23]:
df.shape

(178, 14)

2. Print the column names

In [24]:
df.columns

Index(['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium',
       'total_phenols', 'flavanoids', 'nonflavanoid_phenols',
       'proanthocyanins', 'color_intensity', 'hue',
       'od280/od315_of_diluted_wines', 'proline', 'class'],
      dtype='object')

3. Check the data types

In [25]:
df.dtypes

alcohol                         float64
malic_acid                      float64
ash                             float64
alcalinity_of_ash               float64
magnesium                       float64
total_phenols                   float64
flavanoids                      float64
nonflavanoid_phenols            float64
proanthocyanins                 float64
color_intensity                 float64
hue                             float64
od280/od315_of_diluted_wines    float64
proline                         float64
class                             int64
dtype: object

4. Check missing values

In [28]:
df.isnull().sum()

alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
class                           0
dtype: int64

5. Run describe()

In [29]:
df.describe()

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,class
count,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000
mean,13.000618,2.336348,2.366517,19.494944,99.741573,2.295112,2.029270,0.361854,1.590899,5.058090,0.957449,2.611685,746.893258,0.938202
std,0.811827,1.117146,0.274344,3.339564,14.282484,0.625851,0.998859,0.124453,0.572359,2.318286,0.228572,0.709990,314.907474,0.775035
min,11.030000,0.740000,1.360000,10.600000,70.000000,0.980000,0.340000,0.130000,0.410000,1.280000,0.480000,1.270000,278.000000,0.000000
25%,12.362500,1.602500,2.210000,17.200000,88.000000,1.742500,1.205000,0.270000,1.250000,3.220000,0.782500,1.937500,500.500000,0.000000
50%,13.050000,1.865000,2.360000,19.500000,98.000000,2.355000,2.135000,0.340000,1.555000,4.690000,0.965000,2.780000,673.500000,1.000000
75%,13.677500,3.082500,2.557500,21.500000,107.000000,2.800000,2.875000,0.437500,1.950000,6.200000,1.120000,3.170000,985.000000,2.000000
max,14.830000,5.800000,3.230000,30.000000,162.000000,3.880000,5.080000,0.660000,3.580000,13.000000,1.710000,4.000000,1680.000000,2.000000


6. Write two observations about the data.

(shape of the dataset is 178 rows, 14 columns, without missing values, with 3 classes: divided into 71, 59, 48 rows)
1. Data from describe(): 
 -some of the columns such as, malic_acid, alcanity_of_ash, magnesium, color_intentsity, proline have noticable differences between min, max, and median, and oftenly right-sided distribution. This tells us that for better estimation in future scaling and stardatization will be needed.
2. Data from additional EDA(explaratory data analysis):
- these are the highest corellating columns with positive corellation values, which shows us the importance of these features in future estimation work.
color_intensity                 0.265668
malic_acid                      0.437776
nonflavanoid_phenols            0.489109
alcalinity_of_ash               0.517859
 

In [31]:
df["class"].value_counts()

class
1    71
0    59
2    48
Name: count, dtype: int64

In [33]:
df.corr()["class"].sort_values()

flavanoids                     -0.847498
od280/od315_of_diluted_wines   -0.788230
total_phenols                  -0.719163
proline                        -0.633717
hue                            -0.617369
proanthocyanins                -0.499130
alcohol                        -0.328222
magnesium                      -0.209179
ash                            -0.049643
color_intensity                 0.265668
malic_acid                      0.437776
nonflavanoid_phenols            0.489109
alcalinity_of_ash               0.517859
class                           1.000000
Name: class, dtype: float64

### Student Task Step 3: create X and y

In [34]:
X = df.drop("class", axis = 1)
y = df["class"]

print(X.shape)
print(y.shape)

(178, 13)
(178,)


### Student Task Step 4: split and train

Use Decision Tree again to keep the task simple.

In [36]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 42)

tree = DecisionTreeClassifier(random_state = 42)
tree.fit(X_train, y_train)

prediction = tree.predict(X_test)

### Student Task Step 5: evaluate
They must produce accuracy and a prediction table.

In [41]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, prediction)
print("Accuracy of the prediction is :", accuracy)

results = pd.DataFrame({
    "correct": y_test.values,
    "prediction": prediction
})

results["is_correct"] = results["correct"] == results["prediction"]
display(results.head(10))

Accuracy of the prediction is : 0.9629629629629629


,correct,prediction,is_correct
0,0,0,True
1,0,0,True
2,2,2,True
3,0,0,True
4,1,1,True
5,0,0,True
6,1,1,True
7,2,2,True
8,1,1,True
9,2,2,True


1. feature are columns that needed for training, they are used for estimation; target is column that is used for prediction. Basically, we predict *target* based on *features*
2. df.shape tells us the number of rows and columns in the dataset
3. missing values can affect the training,  for example in the features that we used to train our model missing values can lead to incorrect estimation, as they do not give the model data foundation which is important for prediction. Usually, we replace, handle them by mean values, sometimes median values
4. training data is used for training of the model, for example if we devide the training data into 70/30, 70 percent of the data will be used solely for model to learn a pattern. While, test data, is used for evaluating of the model. Usually models are made to predict real changing values, so before totally using the model in the real environment, we need to test it, run through metrics and see how it learned the pattern
5. accuracy - how well the model performed on the test data, it shows the total correctly predicted target data percentage of all predictions which were made by the model on the given dataset
6. the model is not perfect, mainly because of the dataset, it is small, and the data varies a lot as we could see from EDA, so it i not scaled or standartized. And also we have 3 classes, and we divided the test and train data on random not actually dividing it so that the classes were correctly represented.  However, it shows particularly ideal results, from my point view it can be because of the issues that were discussed previously, which led to over estimation.